### ISSUE TITLE
T*CronModule issues

### ISSUE THREAD

**Issue by szroman** [pradosoft/prado#815]
Hi,

At the beginning I would like to thank you for creating CronModules, nice job.

Issues:
1. New tasks are always executed during the first run of prado-cli cron/run, regardles of scheduled time of execution.
2. Expired tasks (Task0) are being executed every minute after last valid (scheduled) execution.
3. Task01 is scheduled for 15 21 * * * * and after first run next runtime is set one hour later at 22:15:00. My time zone is CET.

Description of process:

1. The way of runnig cron tasks described in comments of TCronModule.php is not clear for me and I am running it this way:

```shell
/usr/bin/sudo -u www-data /usr/bin/php7.4 /home/roman/www/crontestapp/protected/vendor/bin/prado-cli cron/run /home/roman/www/crontestapp/
/usr/bin/sudo -u www-data /usr/bin/php7.4 /home/roman/www/crontestapp/protected/vendor/bin/prado-cli cron/tasks /home/roman/www/crontestapp/
```

2. The way of adding tasks:

```php
$cron = $this->Application->Modules['cron'];

$t = new Prado\Util\Cron\TCronMethodTask('test', 'createTmpFile3("Task0")');
$t->Name ='Task0';
$t->Schedule ='30 10 14 11 * 2022';
$t->UserName ='pracownik4';

$cron->addTask($t);

$t = new Prado\Util\Cron\TCronMethodTask('test', 'createTmpFile3("Task1")');
$t->Name ='Task01';
$t->Schedule ='15 21 * * * *';
$t->UserName ='pracownik4';

$cron->addTask($t);
```

3. My configuration from application.xml:

```xml
<modules>
    ...
    <module id="test" class="Application.Cron.TestCronModule" />
    <module id="cron_db" class="Prado\Data\TDataSourceConfig">
        <database ConnectionString="sqlite:/home/roman/www/crontestapp/protected/Files/cron.jobs" />
    </module>
    <module id="cron" class="Prado\Util\Cron\TDbCronModule" DefaultUserName="pracownik1" ConnectionID="cron_db">
        <--
        <job Schedule="55 11 * * *" Task="test->createTmpFile()" />
        <job Name="cronclean" Schedule="0 0 * * * *" Task="Prado\Util\Cron\TDbCronCleanLogTask" />
        <job Name="dbcacheclean" Schedule="0 0 * * *" Task="cache->flushCacheExpired(true)" />
        -->
    </module>
</modules>
```

4. Tests

BEFORE FIRST RUN
------------------------------------------------------------------
```shell
   Command line tools for Prado 4.2.1.

 Last cron run was 2022-11-14 10:13:35
 The system time is 2022-11-14 10:19:28

   Application Cron Tasks:
 Name   Schedule           Task                          Last Run            Next Run  User        Run #
 Task0  30 10 14 11 * 2022 test->createTmpFile3("Task0") 1970-01-01 01:00:00 01:00:00* @pracownik4 0
 Task01 15 21 * * * *      test->createTmpFile3("Task1") 1970-01-01 01:00:00 01:00:00* @pracownik4 0

 Any 'next run' with a * means it is Pending
```

AFTER FIRST RUN
------------------------------------------------------------------
```shell
  Command line tools for Prado 4.2.1.

Last Task time was 2022-11-14 10:13:35
 Running 2 Cron Tasks @ 2022-11-14 10:20:51

Running Task Task0 as pracownik1
Ending Task Task0

Running Task Task01 as pracownik1
Ending Task Task01

  Command line tools for Prado 4.2.1.

Last cron run was 2022-11-14 10:20:51
The system time is 2022-11-14 10:20:52

  Application Cron Tasks:
Name   Schedule           Task                          Last Run Next Run User        Run #
Task0  30 10 14 11 * 2022 test->createTmpFile3("Task0") 10:20:51 11:30:00 @pracownik4 1
Task01 15 21 * * * *      test->createTmpFile3("Task1") 10:20:51 22:15:00 @pracownik4 1

Any 'next run' with a * means it is Pending
```

AFTER EXPIRATION OF Task0
------------------------------------------------------------------
```shell
  Command line tools for Prado 4.2.1.

Last Task time was 2022-11-14 11:32:01
 Running 1 Cron Tasks @ 2022-11-14 11:33:01 

Running Task Task0 as pracownik1
Ending Task Task0

  Command line tools for Prado 4.2.1.

Last cron run was 2022-11-14 11:33:01
The system time is 2022-11-14 11:33:01

  Application Cron Tasks:
Name   Schedule           Task                          Last Run Next Run  User        Run #
Task0  30 10 14 11 * 2022 test->createTmpFile3("Task0") 11:33:01 01:00:00* @pracownik4 5
Task01 15 21 * * * *      test->createTmpFile3("Task1") 10:20:51 22:15:00  @pracownik4 1

Any 'next run' with a * means it is Pending
```

**Comment by belisoful** [comment]
In step two, what [object and] method are you adding your tasks?

**Comment by belisoful** [comment]
It's looking like the Next Run is getting borked, maybe because of the interval of time or year?

**Comment by szroman** [comment]
From your email:

>I suggest that we add a new boolean variable to cron tasks (for cron) about whether to trigger on the initial timer set or not.
>Should initial trigger be true or false? 

New boolean variable is good solution. I am suggesting false value as default.

> and lastly, not sure what the code bug is yet. very strange that it happens with one and not the other.

It hapens with all tasks, look at step four, Next Run column, AFTER FIRST RUN panel

>It's looking like the Next Run is getting borked, maybe because of the interval of time or year?

It happens with tasks which end at some point in time its schedule and are not needed any more.

>In step two, what [object and] method are you adding your tasks?

Look at step three, application.xml (TestCronModule)
```php
<?php

class TestCronModule extends TModule
{
    public function createTmpFile()
    {
        file_put_contents('/tmp/test77.txt', 'aaa');
    }

    public function createTmpFile2()
    {
        file_put_contents('/tmp/test88.txt', 'bbb');
    }

    public function createTmpFile3($str)
    {
        file_put_contents('/tmp/test99.txt', $str);
    }
}

?>
```

**Comment by belisoful** [comment]
I am agreeing with you on the initial first run shouldn't be, except by parameter.  This is my bad.  Cron is a large module and i made a bad design choice.  I must have been focused on so many other aspects.  I don't think this is a difficult fix.

Also, The lack of a "next date" may be the issue with the Scheduler.  One off tasks was not tested.  sigh.

-> Accepting self imperfections.

**Comment by szroman** [comment]
Learning from mistakes is my way to perfection.

**Comment by belisoful** [comment]
Getting it this point was a massive effort.  These 3 bugs are quite minor in compare.  PRADO needs some kind of standard Cron module.  I've set up "cron" style PRADO tasks quite a bit in my day job.  These modules encode my experiences with it and some expansive creative thinking about it.  My apologies for not checking these obvious cases.  Cron also needed an upgraded CLI.

A unit test is needed.   I'd look at the unit tests for the Schedule parser to make sure that one off dates have proper next dates before and after.  I'm in the process of setting up the tools to get on this if CtrlAltCA doesn't get to it first.  Not sure how quickly I'll be able to get to this. President Not Sure agrees.

This may need a "Prior Date" computation from the schedule parser [like Next Date]...   if there is a prior date, trigger due to being late to that game.  and if there is no prior date, do not trigger on initialize.   Thoughts?

**Comment by belisoful** [comment]
Here:
https://github.com/pradosoft/prado/blob/46bc20d67bfa655de2bbb05128d883c53161c3cf/framework/Util/Cron/TCronTask.php#L195

Change this:
`return time() >= $this->getNextTriggerTime();`

into this:

`
$trigger = $this->getNextTriggerTime();
return $trigger !== null && time() >= $trigger;
`

Does this resolve the repeating cron task triggering after its one off action?

**Comment by belisoful** [comment]
Alright.  I have initial results:

1) triggering tasks on first run seems to be caused by the lack of global task state for LastExecTime. No value is set and so the null value is fed into the TTimeSchedule::getNextTriggerTime.  This faults with a getNextTriggerTime return value of null.   getIsPending isn't checking for null, but assuming it to be "0".  This makes the task "pending" when it's not.

This fixes the initial false trigger in TCronTask
```
public function getNextTriggerTime()
{
	$s = $this->getScheduler();
	return $s->getNextTriggerTime($this->_lastexectime ?? 0);
}
```

2) the faulty recurring task happens when there is no "next run date".  This seems to be caused by TCronTask not processing the null next date properly.  The TCronModuleTest::testGetPendingTasks can easily replicate the bug with the task scheduled: 0 0 1 1 * 2000.

Here is the fix for TCronTask:
```
public function getIsPending()
{
	$trigger = $this->getNextTriggerTime();
	return $trigger !== null && time() >= $trigger;
}
```

It some point I'll make a patch and post a pull request.  But in the mean time, this should correct two of the three issues.


3) The hour being off by one cannot be recreated.  Here is my unit test:

```
	public function testBugReport815()
	{
		//User reports the hour is incremented by by one
		//  eg. 15 21 * * * *  is coming out with a next date of 22:15:00
		$this->assertTime(strtotime("2022-11-13 21:15:00"), '15 21 * * * *', strtotime("2022-11-12 21:16:00"));
		$this->assertTime(strtotime("2022-11-13 21:15:00"), '15 21 * * * *', strtotime("2022-11-12 21:15:01"));
		$this->assertTime(strtotime("2022-11-13 21:15:00"), '15 21 * * * *', strtotime("2022-11-13 21:14:00"));
 
// This next test is proof that the scheduler is doing the next date properly, but that the TCronTask is processing the null date time wrong
	$this->assertTime(strtotime("2022-11-14 10:30:00"), '30 10 14 11 * 2022', strtotime("2022-11-14 10:10:10"));
	$this->assertTime(strtotime("2022-11-14 10:30:00"), '30 10 14 11 * 2022', strtotime("2000-11-14 11:11:11"));
	$this->assertTime(strtotime("2022-11-14 10:30:00"), '30 10 14 11 * 2022', strtotime("1970-01-01 00:00:00"));
		$this->assertNull($this->obj->getNextTriggerTime(strtotime("2022-11-14 10:30:00")));
		$this->assertNull($this->obj->getNextTriggerTime(strtotime("2022-11-14 10:30:01")));
		$this->assertNull($this->obj->getNextTriggerTime(strtotime("2022-11-14 10:31:00")));
		$this->assertNull($this->obj->getNextTriggerTime(strtotime("2022-11-14 10:32:00")));
		$this->assertNull($this->obj->getNextTriggerTime(strtotime("2022-11-14 10:33:00")));
	}
```

This may be an issue with the getNextTriggerTime of TTimeSchedule jumping into UTC.  Is the time zone change to UTC needed?  would that fix the problem?  Is going to UTC the problem?

this needs more looking into.  Timezones are almost universally a "headache" for time based computing.

**Comment by szroman** [comment]
I have applied your patches and look at task 19NovEvery10Minutes. It shouldn't be executed during initialization.
One off task 19Nov2022_10_50 from future is ok.
One off task 18Nov2022_09_50 from past is ok, does not repeat.

```shell
  Command line tools for Prado 4.2.1.

Last Task time was 2022-11-18 11:07:01
 Running 0 Cron Tasks @ 2022-11-18 11:07:58 

  Command line tools for Prado 4.2.1.

Last cron run was 2022-11-18 11:07:58
The system time is 2022-11-18 11:07:58

  Application Cron Tasks: 
Name                Schedule           Task                                     Last Run            Next Run            User        Run # 
19Nov2022_10_50     50 10 19 11 * 2022 test->addAlert("19Nov2022_10_50", 1)     1970-01-01 01:00:00 2022-11-19 11:50:00 @pracownik1 0
19NovEvery10Minutes */10 * 19 11 * *   test->addAlert("19NovEvery10Minutes", 1) 11:04:02            01:00:00            @pracownik1 1
18Nov2022_09_50     50 9 18 11 * 2022  test->addAlert("18Nov2022_09_50", 1)     11:05:02            01:00:00            @pracownik1 1

Any 'next run' with a * means it is Pending
```

**Comment by belisoful** [comment]
We're part of the way there.

a separate issue: I notice that the output of 19Nov2022_10_50, Last run should probably be a '-'.  b/c no last run.
Also the Next run on 18Nov2022_09_50 should be a '-' as well.   b/c no next run.

There is anomaly, indeed. The Unix zero time is showing up as +1 hour for you on the broken "last run" for 19Nov2022_10_50.  not sure what that means.

Let me ask you...  should 18Nov2022_09_50 be triggering?  It hasn't been triggered and is past due at initiation.  This question would determine the behavior of 19NovEvery10Minutes, as there would be a pending task to trigger on initialize.

or should the tasks only be triggered at the strict time of their schedule?  Meaning tasks added into the past won't trigger.

**Comment by szroman** [comment]
>There is anomaly, indeed. The Unix zero time is showing up as +1 hour for you on the broken "last run" for 19Nov2022_10_50. not sure what that means.

It is the same bug (timezone bug). All dates + 1 hour, no exceptions.

>Let me ask you... should 18Nov2022_09_50 be triggering? It hasn't been triggered and is past due at initiation. This question would determine the behavior of 19NovEvery10Minutes, as there would be a pending task to trigger on initialize.

It was executed during initialize and it is ok. Look at run column.

>or should the tasks only be triggered at the strict time of their schedule? Meaning tasks added into the past won't trigger. Or is this the optional parameter?

No! Only as optional parameter.

**Comment by belisoful** [comment]
all the cron times are based upon the php instance time zone rather than the system time zone.  can you determine what time zone your instance is running under?  This may explain the one hour offset.  I have +8 hours between UTC and Pacific Time in my instance. 

I'm looking to simplify.   Let me be 100% clear. You are saying that 18Nov2022_09_50 was right in triggering on initiation but 19NovEvery10Minutes was not right in triggering on initiation despite having a trigger to catch up on?

There is something funky with the next run date for 19NovEvery10Minutes.  I'll look at it after some sleep.

**Comment by szroman** [comment]
>what time zone your instance is running under?

Europe/Berlin

>I'm looking to simplify. Let me be 100% clear. You are saying that 18Nov2022_09_50 was right in triggering on initiation but 19NovEvery10Minutes was not right in triggering on initiation despite having a trigger to catch up on?

Just to be clear. I am writing this on Friday, November 18, 2022, 13:22:56 CET. For me it's November 18th and new tasks with past run time should be run on initialization. Tasks with a future execution time should NOT be run on initialization.

>despite having a trigger to catch up on?

Why do you insist that task 19NovEvery10Minutes is having a trigger to catch up on? Maybe I'm not seeing something?

**Comment by belisoful** [comment]
This makes sense. Got it.  I see what you are saying.  Let me see what cooks up.

Also, The Europe/Berlin PHP timezone explains the +1 hour.  try setting the PHP timezone of the app to UTC or CET, by config settings or in code:

`date_default_timezone_set('UTC');`

or 

`date_default_timezone_set('America/Chicago');`

See if this fixes the +1 hour offset.

**Comment by belisoful** [comment]
I'm adding some time zone info to the cron shell to help see what's up.

**Comment by belisoful** [comment]
My apologies for not seeing that there were no prior scheduled tasks, i didn't see it as the 19 but the 18!

**Comment by belisoful** [comment]
Replace the TCronTask method getNextTriggerTime with this

```
public function getNextTriggerTime()
	{
		$s = $this->getScheduler();
		if ($this->_lastexectime === null) {
			$nowNext = $s->getNextTriggerTime(time() - 1);
			if ($nowNext !== null)
				return $nowNext;
		}
		return $s->getNextTriggerTime($this->_lastexectime ?? 0);
	}
```

I think this is the resolution of the bugs being reported here.   the unit tests need to be updated. that's on the way.

I'll post a pull request when it's uploaded on my fork.

**Comment by belisoful** [comment]
If it is not run EXACTLY on the minute and second, it does not trigger the first time.  This isn't proper behavior.

This is an incomplete solution, but it'll get you kind of the way there.  Still working on it.

**Comment by belisoful** [comment]
T*CronModule bug fix initializes the lastExecTime in global state based upon if there is a next trigger.  It works just as you have specified, correctly.

Thank you for sticking with it and explaining.

Here is the fix. The Test cases are updated and all passing with this code fix.   Soon to be on my fork with a pull request to main.  You should be able to run this and work as assumed.

TCronTask
```
	public function setLastExecTime($v)
	{
		if ($v !== null) {
			$this->_lastexectime = TPropertyValue::ensureInteger($v);
		} else {
			$this->_lastexectime = null;
		}
	}

	public function getNextTriggerTime()
	{
		$s = $this->getScheduler();
		return $s->getNextTriggerTime($this->_lastexectime ?? 0);
	}

	public function getIsPending()
	{
		$trigger = $this->getNextTriggerTime();
		return $trigger !== null && time() >= $trigger;
	}
```

TCronModule
```
	protected function setPersistentData($name, $task)
	{
		$tasksInfo = $this->getApplication()->getGlobalState(self::TASKS_INFO, []);
		if (isset($tasksInfo[$name])) {
			$task->setLastExecTime($tasksInfo[$name]['lastExecTime']);
			$task->setProcessCount($tasksInfo[$name]['processCount']);
			return true;
		} else {
			$now = time();
			$schedule = $task->getScheduler();
			$nextTriggerTime = $schedule->getNextTriggerTime($now);
			$tasksInfo[$name]['lastExecTime'] = ($nextTriggerTime === null) ? null : (($schedule->getNextTriggerTime(0) == $nextTriggerTime) ? null : $now);
			$tasksInfo[$name]['processCount'] = 0;
			$this->getApplication()->setGlobalState(self::TASKS_INFO, $tasksInfo, []);
			$task->setLastExecTime($tasksInfo[$name]['lastExecTime']);
			$task->setProcessCount($tasksInfo[$name]['processCount']);
		}
		return false;
	}
```

TDbCronModule
```
	protected function setPersistentData($name, $task)
	{
		if (isset($this->_taskRows[$name])) {
			$time = $task->getLastExecTime();
			$count = $task->getProcessCount();
			$task->setLastExecTime($this->_taskRows[$name]['lastexectime']);
			$task->setProcessCount((int) $this->_taskRows[$name]['processcount']);
			if (serialize($task) !== $this->_taskRows[$name]['options']) {
				$task->setLastExecTime($time);
				$task->setProcessCount($count);
				$this->updateTaskInternal($task);
				return false;
			}
			return true;
		} else {
			$now = time();
			$schedule = $task->getScheduler();
			$nextTriggerTime = $schedule->getNextTriggerTime($now);
			$lastExecTime = ($nextTriggerTime === null) ? null : (($schedule->getNextTriggerTime(0) == $nextTriggerTime) ? null : $now);
			$task->setLastExecTime($lastExecTime);
			$task->setProcessCount(0);
			$this->addTaskInternal($task);
		}
		return false;
	}
```

**Comment by szroman** [comment]
I have applied your patches. Task 24Nov10past10 should not be triggered on init and timezone issue still exists.

```shell
  Command line tools for Prado 4.2.1.

Last Task time was 2022-11-21 14:26:01
 Running 1 Cron Tasks @ 2022-11-21 14:26:19 

Running Task 24Nov10past10 as pracownik1
Ending Task 24Nov10past10

  Command line tools for Prado 4.2.1.

Last cron run was 2022-11-21 14:26:19
The system time is 2022-11-21 14:26:20

  Application Cron Tasks: 
Name          Schedule        Task                               Last Run Next Run            User        Run # 
24Nov10past10 10 10 24 11 * * test->addAlert("24Nov10past10", 1) 14:26:20 2022-11-24 11:10:00 @pracownik1 1

Any 'next run' with a * means it is Pending 
```

**Comment by szroman** [comment]
With this version of getNextTriggerTime doesn't trigger on init, only timezone issue remains.

```php
/**
	 * @return numeric time of the next trigger after the lastExecTime
	 */
	public function getNextTriggerTime()
	{
		$s = $this->getScheduler();
		if ($this->_lastexectime === null) {
			$nowNext = $s->getNextTriggerTime(time() - 2);
			if ($nowNext !== null)
				return $nowNext;
		}
		return $s->getNextTriggerTime($this->_lastexectime ?? 0);
	}
```

**Comment by belisoful** [comment]
man.  I thought i tested that kind of condition, and it was working.  Alright.  Testing continues with a new schedule.  Are you using regular or DB cron?  

the 1 hour difference can be explained by your PHP timezone being set to +1:00.  Try setting your PHP Application timezone to "UTC" and i bet the difference will go away. like this in your application.xml:

```
 <module class="Prado\Util\TBehaviorsModule" >
     <behavior name="TimeZoneParameter" Class="Prado\Util\Behaviors\TTimeZoneParameterBehavior" AttachTo="Application" TimeZoneParameter="" TimeZone="UTC"/>
  </module>
```

**Comment by belisoful** [comment]
I popped in your schedule and it was correctly NOT running on initialization.  Meaning, the changes are working on my end.  I cannot explain it.  I do have some broader changes like to the test units and cron shell output (nulls in lastExecTime showing as "never" rather than 1970-1-1).

I've been looking at the timezone and you may be right about the schedule running in UTC only.  Running them in the specified time zone may be the behavior you are expecting, yes?  That's what I'm attempting to wrap my head around...  and the solution.

**Comment by belisoful** [comment]
I am unable to recreate the problem you are having [with running on init when it shouldn't] (with the new code).  

I can recreate a task initialization [and execution on init timing] bug when an app.xml task schedule is changed without changing the name of the task in DBCron.  Like if the task schedule is */10 and is changed to */5 without a name change.  Changing the name of the task is automatic in normal cron because without a name it hashes the schedule and task for name and hence differences make different task names (and reset the tasks), but DB Cron seems to be having issues with changing the schedule without the name change.  Looking into it now.

Normal Cron can change schedule without name change, but the last run is not reset with a changed schedule.  So, if a schedule is reset to only trigger before the last run time, nothing is every executed.  The name keeps the old lastexectime.

Changing the schedule of a named task, as of right now, is not supported.  Any changes to a schedule needs a name change too.  This may be confusing for developers.  We may need a hash check (stored in global state [and db?]) on the name-schedule-task and reset the lastExecTime on any changes.  Thoughts anyone?

Also, I think I figured out the timezone issue, the getNextTriggerTime shouldn't be normalizing the timezone to UTC.  I'm removing that now and it seems to resolve the issue.

In TTimeScheduler::getNextTriggerTime
comment out these lines
```
#547 		//$restoreZone = date_default_timezone_get();
#548		//date_default_timezone_set("UTC");
...
#666		//date_default_timezone_set($restoreZone);
...
#696		//date_default_timezone_set($restoreZone);
```

**Comment by belisoful** [comment]
Alright.  I have a fix for a more graceful failure of DBCron when changing the schedule without reseting the name.  This should stop the run on init when changing the schedule of an existing job of DB Cron.

TDbCronModule change.
```
	protected function setPersistentData($name, $task)
	{
		if (isset($this->_taskRows[$name])) {
			$time = $task->getLastExecTime();
			$task->setLastExecTime($this->_taskRows[$name]['lastexectime']);
			$task->setProcessCount((int) $this->_taskRows[$name]['processcount']);
			if (serialize($task) !== $this->_taskRows[$name]['options']) {
				$now = time();
				$schedule = $task->getScheduler();
				$nextTriggerTime = $schedule->getNextTriggerTime($now);
				$lastExecTime = ($nextTriggerTime === null) ? null : (($schedule->getNextTriggerTime(0) == $nextTriggerTime) ? null : $now);
				$task->setLastExecTime($lastExecTime);
				$this->updateTaskInternal($task);
				return false;
			}
			return true;
		} else {
			$now = time();
			$schedule = $task->getScheduler();
			$nextTriggerTime = $schedule->getNextTriggerTime($now);
			$lastExecTime = ($nextTriggerTime === null) ? null : (($schedule->getNextTriggerTime(0) == $nextTriggerTime) ? null : $now);
			$task->setLastExecTime($lastExecTime);
			$this->addTaskInternal($task);
		}
		return false;
	}
```

**Comment by belisoful** [comment]
I'm cleaning it up a bit, and hopefully will be posting this ASAP with pull request.  Cron is very important to get right.

**Comment by szroman** [comment]
Changing timezone to UTC is not an option. Different users use different timezones in my apps.

I commented out one line in the TTimeScheduler (as you suggested) that changed (for good) the way NextRun is calculated.

```php
$restoreZone = date_default_timezone_get();
// date_default_timezone_set("UTC");
if ($priortime === false) {
```

BEFORE INIT

1.All LastRun should be 1970-01-01 00:00:00, but it is not so importent.
2.Task Future NextRun value is incorrect and pending. It should be exactly as in FutureWithYear.

```shell
Application Cron Tasks: 
Name           Schedule           Task                                Last Run            Next Run            User        Run # 
PastWithYear   10 10 20 11 * 2022 test->addAlert("PastWithYear", 1)   1970-01-01 01:00:00 10:10:00*           @pracownik1 0
Past           10 10 20 11 * *    test->addAlert("Past", 1)           1970-01-01 01:00:00 10:10:00*           @pracownik1 0
FutureWithYear 10 10 24 11 * 2022 test->addAlert("FutureWithYear", 1) 1970-01-01 01:00:00 2022-11-24 10:10:00 @pracownik1 0
Future         10 10 24 11 * *    test->addAlert("Future", 1)         1970-01-01 01:00:00 10:10:00*           @pracownik1 0
```

AFTER INIT

```shell
Application Cron Tasks: 
Name           Schedule           Task                                Last Run            Next Run            User        Run # 
PastWithYear   10 10 20 11 * 2022 test->addAlert("PastWithYear", 1)   09:56:45            01:00:00            @pracownik1 1
Past           10 10 20 11 * *    test->addAlert("Past", 1)           09:56:46            2023-11-20 10:10:00 @pracownik1 1
FutureWithYear 10 10 24 11 * 2022 test->addAlert("FutureWithYear", 1) 1970-01-01 01:00:00 2022-11-24 10:10:00 @pracownik1 0
Future         10 10 24 11 * *    test->addAlert("Future", 1)         09:56:47            2022-11-24 10:10:00 @pracownik1 1
```

**Comment by belisoful** [comment]
Your input is invaluable. 🙏 Thank you.

**Comment by belisoful** [comment]
Are you using normal cron or DBCron?

I tested both and then edited both without changing the name.   It did not replicate the behavior you are seeing.  It was working properly for me.

The "last run" for Past and Future in your BEFORE INIT examples should be the first load time so that the NextRun is properly the first next scheduled time rather than a prior time..

glad the timezone issue is resolved by removing code.

**Comment by szroman** [comment]
It is described in the first comment of this thread (TDbCronModule, TCronMethodTask) or are you asking about linux cron version?

**Comment by belisoful** [comment]
I was asking if you had changed your configuration.  There are many "options" for doing these things.  Cron is a very open module.  It can do one thing many different ways.  I got your email.  I'll send the files over shortly.

**Comment by belisoful** [comment]
Tip, instead of creating test methods, you can use dynamic events.  like PRADO "on" prefix for object events, the new prefix "dy" is a method space that, by default, always exists.  So all "dy" events can be called, all the time.  It's a matter of an object implementing it... and more specifically, a behavior or a class behavior implementing it for the owner object.

I say this to simplify your testing.  You can call any application module and any dynamic event.  so the "cron" method task could be "cron->dyAnything('default return value', 'any param', 3, 6.9, true, false)".  all dynamic events will do nothing and return false.  It's a passthrough to the behaviors' implementation.  I've been using "dy" events to test, fyi.  full disclosure for bug reporting.

That shouldn't make a difference in task scheduling and testing.

There is a global event paradigm using the new "fx" event, but that isn't useful here.  It's for advanced module development and cross module development.  It's specifically made for interactions between public modules installed via composer without knowledge of each other.  It's a brainiac remake of Wordpress global filter and "event" system that PRADO should probably have in the long run.

composer modules can be specified to instance in the app.xml much more easily.  it's a tighter integration, like that other platforms had, eg Yii, and PRADO needed.  I'm not seeing many composer modules being developed for PRADO yet.

**Comment by szroman** [comment]
Thanks for all the tips. You are creating something amazing for others. Don't apologize for mistakes. Mistakes make the road more interesting.

Setting LastExecTime solved the problem of starting future tasks during initialization.

```php
$cron = $this->Application->Modules['cron'];

$t = new Prado\Util\Cron\TCronMethodTask('test', 'addAlert("Test1", 1)');
$t->Name = "Test1";
$t->Schedule = "10 10 24 11 * *";
$t->UserName = "pracownik1";
$t->LastExecTime = time(); //OBLIGATORY !!!

$cron->addTask($t);
```

This way of adding works fine. For those who want to use TDbCronModule, remember to set the UserManager parameter to ensure proper login.

```xml
<module id="users2" class="Application.Common.MyUserManager" />
<module id="test" dbname="pgsql" class="Application.Cron.TestCronModule" />
<module id="cron_db" class="Prado\Data\TDataSourceConfig">
    <database ConnectionString="sqlite:/home/roman/www/devel/protected/Files/cron.jobs" />
</module>
<module id="cron" class="Prado\Util\Cron\TDbCronModule" DefaultUserName="pracownik1" ConnectionID="cron_db" UserManager="users2">
    <job Name="Test0" Schedule="10 10 24 11 * *" Task="test->addAlert('Test0', 1)" UserName="pracownik1" />
</module>
```

**Comment by belisoful** [comment]
> Setting LastExecTime solved the problem of starting future tasks during initialization
This will deactivate any unrun tasks that are scheduled before the lastExecTime.  Meaning unrun tasks scheduled before they install won't run.

If you want to preset LastExecTime, this is the ideal computation:
```
$now = time();
$schedule = $task->getScheduler();
$nextTriggerTime = $schedule->getNextTriggerTime($now);
$lastExecTime = ($nextTriggerTime === null) ? null : (($schedule->getNextTriggerTime(0) == $nextTriggerTime) ? null : $now);
$task->setLastExecTime($lastExecTime);
```

The code update [ready to commit?] uses this to initialize lastExecTime on a new task or a changed db task.  When the task is not found in the global state, or db [for the DBCron, respectively], lastExecTime is initialized with the above code.  I'm still noodling on it and any other updates.

Setting the UserManager isn't strictly enforced, but it is certainly performant.  Cron makes use of TApplication::getModulesByType to search for the IUserManager module in the application as a fall back.  Make sure you have an IUserManager module in the app for the cron task user[s], if you use the task user[s].

**Comment by szroman** [comment]
Just to be clear. Look at debug info.
--------------------------------------

Setting LastExecTime in TCronTask object is necessary only for task meant to be triggered in the future.
Why did the TestXml task calculate LastRun but the TestCronMethodTask task did not?
Compare NextRun columns. All ways of adding should have the same end result.

```xml
<module id="cron" class="Prado\Util\Cron\TDbCronModule" DefaultUserName="pracownik1" ConnectionID="cron_db" UserManager="users2">
    <job Name="TestXml" Schedule="10 10 28 11 * *" Task="test->addAlert('TestXml', 1)" UserName="pracownik1" />
</module>
```

```php
$cron = $this->Application->Modules['cron'];

$cron->removeTask("TestCronMethodTask");

$t = new Prado\Util\Cron\TCronMethodTask('test', 'addAlert("TestXml", 1)');
$t->Name = "TestCronMethodTask";
$t->Schedule = "10 10 28 11 * *";
$t->UserName = "pracownik1";
// $t->LastExecTime = time();

$cron->addTask($t);
```

BEFORE INIT
```shell
Application Cron Tasks: 
Name               Schedule        Task                                    Last Run     Next Run             User        Run # 
TestCronMethodTask 10 10 28 11 * * test->addAlert("TestCronMethodTask", 1) -            1970-11-28 10:10:00* @pracownik1 0
TestXml            10 10 28 11 * * test->addAlert('TestXml', 1)            - (13:10:33) 2022-11-28 10:10:00  @pracownik1 0
```

AFTER INIT
```shell
Application Cron Tasks: 
Name               Schedule        Task                                    Last Run     Next Run            User        Run # 
TestCronMethodTask 10 10 28 11 * * test->addAlert("TestCronMethodTask", 1) 13:22:31     2022-11-28 10:10:00 @pracownik1 1
TestXml            10 10 28 11 * * test->addAlert('TestXml', 1)            - (13:10:33) 2022-11-28 10:10:00 @pracownik1 0
```
WITH LastExecTime
-----------------------------------------------------

```php
$cron = $this->Application->Modules['cron'];

$cron->removeTask("TestCronMethodTask");

$t = new Prado\Util\Cron\TCronMethodTask('test', 'addAlert("TestXml", 1)');
$t->Name = "TestCronMethodTask";
$t->Schedule = "10 10 28 11 * *";
$t->UserName = "pracownik1";
$t->LastExecTime = time();

$cron->addTask($t);
```

BEFORE INIT
```shell
Application Cron Tasks: 
Name               Schedule        Task                                    Last Run     Next Run            User        Run # 
TestCronMethodTask 10 10 28 11 * * test->addAlert("TestCronMethodTask", 1) - (13:23:33) 2022-11-28 10:10:00 @pracownik1 0
TestXml            10 10 28 11 * * test->addAlert('TestXml', 1)            - (13:10:33) 2022-11-28 10:10:00 @pracownik1 0
```

AFTER INIT
```shell
Application Cron Tasks: 
Name               Schedule        Task                                    Last Run     Next Run            User        Run # 
TestCronMethodTask 10 10 28 11 * * test->addAlert("TestCronMethodTask", 1) - (13:23:33) 2022-11-28 10:10:00 @pracownik1 0
TestXml            10 10 28 11 * * test->addAlert('TestXml', 1)            - (13:10:33) 2022-11-28 10:10:00 @pracownik1 0
```

**Comment by belisoful** [comment]
I think it's missing the lastExecTime reset on that [last] way of adding tasks.  That's my oversight, again.  I'll add that tomorrow and send forth an update to you via email.

and again, thank you for your extremely thorough testing!!!

In the spirit of progress, I am looking at the TTimeScheduler for optimizations.  The regex is being recomputed each time and can be cached.  Also, The crontab schedule validation, parsing, and nextTriggerDate computation is "monumental", relatively speaking.

So, I am also thinking of adding a hyper efficient one off time scheduler format...  where a unix time is specified like "@1669209253."  This bypasses the complex schedule validator/parser and the next trigger time computations.  Test unit need to be written.

This should allow the system to handle many more one-off user based cron tasks because the schedule parsing and next trigger compute is minimized.  Thoughts?

**Comment by belisoful** [comment]
Also, I'm looking at the DBCron and thinking...  why are all the tasks loaded to check them for pending?  How about a DB field for NextRun based on lastExecTime, and then only parsing out those objects where nextRun is before now?  Let the DB to the lifting for checking DBCron time conditions, it would be far more efficient.

I think this can be optimized more so that adding, updating, getPending, and getTask isn't loading all the tasks for their work.  Looking into it....

**Comment by belisoful** [comment]
Almost done with a refactor of DBCron to lazy load only the necessary tasks, eg. pending, adding, updating, removing.  This will allow for heavy user service usage of cron.  It adds a "nextexectime" and two INDEX for the username and nextexectime to the database.  New functionality for paged GetTasks and new functions for paged getUserTasks, and getUserTaskCount.

**Comment by belisoful** [comment]
I have v2 of DB cron ready.  Should this pull request be accepted before v2 is submitted?

**Comment by szroman** [comment]
I love the shortcuts you introduced. Please give me two days to test the changes you have made.

**Comment by belisoful** [comment]
Thank you.  An architect is as an architect does.  It only makes sense.  I was reading the energy of the room.

The unit tests in v2 will remain the same in what you are testing... meaning all the lastexecdate initializations for various schedules are retained by the new v2 code exactly.  Testing will be very useful with this and the v2 code.

Of course, the usage of an exact seconds schedule means that second accuracy can be created.  What are peoples' thoughts on having an option for not just minute accuracy but seconds accuracy?

My thoughts on how to implement:
When Cron Module SecondsAccuracy is true (default:false)
Get the pending tasks to 59th second of the Current minute.
order the pending tasks by nextTriggerTime ascending
while there are pending tasks to run and not after the current minute {
     run tasks scheduled prior or equal to now
     sleep until next pending task is scheduled.
}

My only concern is potential for double running tasks at the later in the minute tasks that take longer than the current minute to compute.  Maybe there should be a "current running task" in the global state to prevent double task runs in SecondsAccuracy mode.

**Comment by belisoful** [comment]
I have a TCronModule setting for SecondResolution, default false; coming in v2.  It inserts process sleeping to delay task trigger when the task is to trigger within the minute.  This provides "to the second" task triggering made available with the  "at time" efficient one-off schedule format.

Global State won't provide the mutex to a process that runs at, say, the 59th second and takes longer than that minute to execute, in regards to the next cron process executing the same task.  Using Global State only adds further complexities that aren't resolved with global state.  So, it goes simple.

If the cron process is running long (into the next cron process) in secondResolution mode, it'll stop processing cron tasks and let the next cron process the pending tasks.

**Comment by szroman** [comment]
If the task takes 5 minutes, it will trigger 5 times (pending == true) until the first trigger finishes (pending == false). How to deal with this? Is this correct functionality?

**Comment by belisoful** [comment]
One off tasks should only trigger once.  Updating the lastexectime happens after the task runs.  Updating the lastexectime before the task runs would avoid this and stop last second double runs.  Your suggestion seems better because the task takes an unknown amount of time, and may even time out.  Agreed.

recurring tasks that take longer than recur time should retrigger.

**Comment by belisoful** [comment]
The reordering of updating the lastExecTime from after to before tasks run.  Tasks may run for an undetermined amount of time.

**Comment by szroman** [comment]
Still something not working. Each task takes 300 seconds to complete. Looks nice, but actually 10 Alerts have been added. Look at LastRun column.

```shell  Application Cron Tasks: 
Name    Schedule           Task                           Last Run Next Run            User        Run # 
Rec0    56 14 30 11 * *    test->addAlert("Rec0", 300)    14:56:01 2023-11-30 14:56:00 @pracownik1 1
Rec1    56 14 30 11 * *    test->addAlert("Rec1", 300)    15:01:01 2023-11-30 14:56:00 @pracownik1 1
OneOff0 56 14 30 11 * 2022 test->addAlert("OneOff0", 300) 15:06:01 -                   @pracownik1 1
OneOff1 56 14 30 11 * 2022 test->addAlert("OneOff1", 300) 15:11:01 -                   @pracownik1 1
```
From my test app:
![image](https://user-images.githubusercontent.com/18717342/204821489-45101b8d-7ae2-49a3-aa17-0c5330b2e57c.png)

**Comment by belisoful** [comment]
It's looking like it is properly delaying 300 seconds on each task.  But the logic for stopping current cron when running into the next minute cron (on LONG tasks) isn't in v1 yet.  This is taken care of in v2.  I'll port over the logic from v2 to v1 shortly and we'll see how that works.

It is properly executing Rec1, OneOff0, and OneOff1 in separate cron task minute processes due to being blocked by Rec0, etc in the original executing tasks.  I've looked at threading and parallelizing tasks for each time of pending tasks, but shutting down one task thread shuts down the resources for all the threads.  A PHP multithreading expert may be useful.  Could the garbage collection of each multithreaded task be delayed until they all end?  Memory implications?

**Comment by belisoful** [comment]
An update is posted. but it's not the final solution.  If running cron run manually at 59.99 seconds, it'll refuse to run any tasks when it gets to the pending tasks (where the first time is the prior minute and task time is current minute).  Such time dependency for untimed manually running tasks is (likely) unacceptable.  If the cron tab is run manually, it should run all the way.  Trading one bug for another.

Should there be maybe a CLI option, like '--cronMinute', to turn on the logic for handling the long-multi-run problem from multiple running cron tasks?  There needs to be a distinction between manual cron run and system cron run.

Looking into it....

**Comment by belisoful** [comment]
Are there any shell environment variables set by cron only seen by cron run processes that we can select the "pending tasks only in the current cron-minute" logic upon?

**Comment by belisoful** [comment]
Here are the two outputs from the shell environments, terminal and cron.

CLI has this shell environment:
```
TERM_PROGRAM=Apple_Terminal
SHELL=/bin/zsh
TERM=xterm-256color
TMPDIR=/var/folders/ry/teqh84395hix2cqklupmb2toc/T/
TERM_PROGRAM_VERSION=442
TERM_SESSION_ID=84736519-BDAA-DCDC-8D8D-00CC4738761B
USER=user
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.hutc7ud34/Listeners
PATH=/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
LaunchInstanceID=FCB65340-D5B9-84A2-E487-47CAF472689 
__CFBundleIdentifier=com.apple.Terminal
PWD=/Users/user/Code/app/protected/vendor/pradosoft/prado
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
SHLVL=1
HOME=/Users/user
LOGNAME=user
SECURITYSESSIONID=246cc
OLDPWD=/Users/user/Code/app/protected/vendor/pradosoft/prado
LANG=en_US.UTF-8
_=/usr/bin/env
```

Cron has this shell environment:
```
SHELL=/bin/sh
USER=user
PATH=/usr/bin:/bin
PWD=/Users/user
SHLVL=1
HOME=/Users/user
LOGNAME=user
_=/usr/bin/env
```

The difference is staggering.

in reading about it, cron "usually" has SHELL=/bin/sh and PATH=/usr/bin:/bin.  This can be a selector for turning on the cron specific minute based logic. default false.

Conversely, these are selectors for turning off the cron specific minute based logic: LANG, TERM, and TERM_PROGRAM. default true.

Any thoughts?  I am leaning to turning off the cron specific minute based logic for LANG or TERM or TERM_PROGRAM for proper behavior of the cron/run between environments.

**Comment by belisoful** [comment]
I have the "pending tasks in the current minute-only" logic running based upon CLI being cron (vs Termina)l; for cron and not in Terminal mode.  Terminal runs all the tasks regardless of timing as expected. It's updated.

Running multiple Prado cron manually from terminal will cause problems in running erroneously, especially for long run tasks, but the issue is mitigated in multiple crontab runs.  I doubt that users would manually run Prado cron multiple times and/or while running system crontab to run Prado cron.

Having a global variable to say which tasks are pending would resolve this but in Cron, the global state doesn't have the resolution needed for mutexing pending tasks from multiple cron processes.  DBCron could do it with an added DB column for "pending."  Coordinating the pending would be a nightmare and error conditions could prevent tasks from running ever.

I think the behavior of T*Cron from crontab vs terminal was confusing matters and why the bug was not initially identified.

My questions to the community are about RequestCron.
1) should set_time_limit be changed in RequestCron?
2) should each task set_time_limit in RequestCron?
3) What about the web server time limit?
4) what about tasks that expire from the time limit?

**Comment by belisoful** [comment]
The latest code passed all tests except the phpstan mysql upgrade?

**Comment by belisoful** [comment]
The php @method was corrected (but not linked here).  All the tests passed except upgrading the mysql database?

**Comment by belisoful** [comment]
@ctrlaltca If I recall, you are on developing and testing on MS Windows.  How does one detect an automated cron script in MS Windows (vs linux/macOS/debian)?  Can we build such detection into TShellApplication for MS Windows?

In Debian and MacOS, from the crontab, Cron and cron-sudo is showing a lack of $_ENV['TERM'] or is equal to 'unknown', respectively.  What kind of automated cron/crontab detection can be built in for Windows?   Or is a command line option needed for that?

Adding a command line option is easy to register with the latest upgrades to Shell.

**Comment by ctrlaltca** [comment]
> If I recall, you are on developing and testing on MS Windows

Oh, the horror. No i'm sorry i'm on linux mostly

I guess you can use some powershell voodoo to query scheduled tasks, this looks like a good starting point:
https://learn.microsoft.com/en-us/powershell/module/scheduledtasks/get-scheduledtask?view=windowsserver2022-ps

**Comment by belisoful** [comment]
Belief in Windows must be something akin to belief in the spaghetti noodle monster...  ?  Glad to be wrong about that.  sorry to bother you.

I think someone wise in the MS ways must be consulted...   Is there anyone familiar with Powershell cron? anywhere?  willing to help?  Developers in the user base?  Powershell looks like a mess.

Am I asking into the void of nothingness?

I have already added a "--cron" CLI option for our Windows users.  It limits tasks being executed to the current minute.  If it's not set, it'll auto-detect and work for what seem to be most linux users.  We know the next minute is going to re-execute, so if a task is taking a long time, any buffered pending tasks will not execute due to being processed by the next cron minute.

I think the Cron v1 is nearing completion.  Cron v2 will be oriented to being cron for users as well and incorporate the v1 update.

There are a few edit-updates to the Shell required to work just right.  tiding up things.  I'll be coming shortly.

**Comment by belisoful** [comment]
Roman said this in his last email to me:

> It works! My testing is nothing compared to what you guys create. Thank you.

**Comment by belisoful** [comment]
These bugs are squashed.  Time to close this issue.  Thank you for reporting and testing. This also resulted in some tweaks to cron making the Db more "per user" friendly.  that's cron v2 being posted shortly.
